### ISSUE TITLE
Node version (10) End of Life

### ISSUE THREAD

**Issue by oguzkilcan** [charleszipp/azure-pipelines-tasks-terraform#447]
I have started to receive a warnings on the Node version being end-of-life.

The actual warning message i get is shown below….

##[warning]Task 'Terraform Installer' version 1 (TerraformInstaller@1) is dependent on a Node version (10) that is end-of-life. Contact the extension owner for an updated version of the task. Task maintainers should review Node upgrade guidance: https://aka.ms/node-runner-guidance

##[warning]Task 'Terraform CLI' version 1 (TerraformCLI@1) is dependent on a Node version (10) that is end-of-life. Contact the extension owner for an updated version of the task. Task maintainers should review Node upgrade guidance: https://aka.ms/node-runner-guidance

I believe a small code block inside `task.json` files is the reason of these warnings:
tasks/terraform-installer/task.json
tasks/terraform-cli/task.json
```    
"execution": {
        "Node10": {
            "target": ".bin/index.js",
            "argumentFormat": ""
        }
    }```

**Comment by JAK1047** [comment]
Same here.

**Comment by UdayMadamsetty** [comment]
Warnings are being reported on Azure DevOps pipeline, even though updated to latest available task version @2. Provided screenshot below for your reference. Could you please help us understand its impact on pipeline runs using these tasks? We are worried on its impact considering deployments to multiple environments through pipelines that are based on these tasks.

![image](https://github.com/user-attachments/assets/18a0d039-186a-4da0-94ef-addf00b4087a)

**Comment by andrei-ivanovici** [comment]
Are there any actual hard dependencies on node 10 ? Or is it just a matter of dependencies being out of date ?

**Comment by Noel-Jones** [comment]
It would appear to be a hard dependency set here https://github.com/jason-johnson/azure-pipelines-tasks-terraform/blob/2d2bf62185e002951bcc370c03efb107fc4c912a/tasks/terraform-installer/task.json#L39 and https://github.com/jason-johnson/azure-pipelines-tasks-terraform/blob/2d2bf62185e002951bcc370c03efb107fc4c912a/tasks/terraform-cli/task.json#L477

**Comment by joezuniga19** [comment]
Great news this is reported.  I initially got this error using the older Charles Zapp version.  I reinstalled the new extension from Jason Johnson, and imagine my surprise when the error was still present.  I look forward to the progress on this report.

**Comment by PratMoha** [comment]
Same issue affecting most of our pipelines as well with Node version 10 end of life.

@jason-johnson  @charleszipp

**Comment by jason-johnson** [comment]
For pull request on this issue, it will need to start by checking out the 1. Tag so that the change will back port. 2. Is not considered stable and I think 0. Is actually broken.

**Comment by JAK1047** [comment]
> For pull request on this issue, it will need to start by checking out the 1. Tag so that the change will back port. 2. Is not considered stable and I think 0. Is actually broken.

I know I'm currently using the @1 tag myself. Are you saying a new version will be pushed with the node version change so long as you are using the @1 tag?

**Comment by cveld** [comment]
@JAK1047 well @jason-johnson is encouraging us to create a PR to fix this node10 dependency issue and point us to use the @1 tag as the base.

**Comment by Noel-Jones** [comment]
Looking at the documentation for the extension it says "For example, currently TerraformCLI@0 and TerraformInstaller@0 will point to the version with the tag 0.7.11. TerraformCLI@1 and TerraformInstaller@1 will point to the version with the tag 1.1.0 and TerraformCLI@2 and TerraformInstaller@2 will point to the unstable version, basically the latest version which has been published to the main branch of the project." Am I to understand from this and the comments that we are talking about a "hotfix" based on on tag 1.1.0? I don't think 1.1.1 was published (particularly since the next commit says "fix publish") and there have been a fair number of commits to main since then - presumably the unstable bit. I'm not sure how such a hotfix would be committed and published.

**Comment by fardarter** [comment]
@jason-johnson What would be the target for the PR? I don't see any maintenance branches for different tag major versions.

**Comment by jason-johnson** [comment]
There are no need for maintenance branches.  With git (n case someone reading this doesn't know) you can check out tags so tags double as "maintenance branches" without cluttering up the repo.  When I make fixes like this, I check out the tag, make a branch from that commit, make the fix and then do the PR.

In open source projects, it's great when we can get contributions.  If no one makes a PR I will eventually do it but it will often happen faster with a PR.  I tried to word the above to help anyone who might make a PR for this, including me. :)

**Comment by fardarter** [comment]
@jason-johnson I understand how to check out a tag. What I'm asking is what the target branch for the PR is supposed to be once changes have been added. 

Unless you're maintaining the `main` branch as the v1 branch and the change just goes on top of `main`, you would need maintenance branches for other tag lineages. Otherwise, how would you work on a v2 and still have the ability to add new code just to the v1 lineage?

**Comment by jaalexan1983** [comment]
@jason-johnson any movement on this PR for node version?

**Comment by jason-johnson** [comment]
> @jason-johnson I understand how to check out a tag. What I'm asking is what the target branch for the PR is supposed to be once changes have been added.
> 
> Unless you're maintaining the `main` branch as the v1 branch and the change just goes on top of `main`, you would need maintenance branches for other tag lineages. Otherwise, how would you work on a v2 and still have the ability to add new code just to the v1 lineage?

I explicitly said "(for those who don't know)" so as not to assume what anyone specifically knew or didn't know about how git works. :)

In any case, right now there is the old "0" version.  Then there is the "1" version which is tagged off of the main branch.  The "2" version is actually just what ever has been merged into main and not tagged.  This way we can quickly check fixes without affecting older releases.

**Comment by fardarter** [comment]
@jason-johnson So the implication here is that I need to use 2 to get the fix, even though it's potentially unstable because of other stuff that has gone in between 1 and 2?

**Comment by jaalexan1983** [comment]
@jason-johnson thanks for merging in, i just tried version 2 of the cli and i am still getting the node 10 warning.  When do these changes get replicated to the market place? https://marketplace.visualstudio.com/items?itemName=JasonBJohnson.azure-pipelines-tasks-terraform

**Comment by jason-johnson** [comment]
I believe the warning is happening because I have not yet tagged v1.  The way ADO extensions work (for those who don't know) is that there is actually a whole copy of the extension for each "version" so to fix this we may have to cherry pick the fix to all 3 "versions" instead of just 1 and 2.

**Comment by jaalexan1983** [comment]
@jason-johnson thank you for the update.  How long do you anticipate these changes will make it top the Market Place, is it an automatic process once you tagged to v1?  I noticed in my extensions list that the last update for terraform tasks was 18th March.

**Comment by jason-johnson** [comment]
It would happen within minutes.  I haven't tagged because I'm unable to test version 2 right now for some reason.  I need to get that solved before I push this to version 1.

**Comment by favoretti** [comment]
@jason-johnson for whatever reason tfx fails to publish from a silicon mac, it crashes with some odd errors, otherwise I'd have tested it in my private publishery...

**Comment by jason-johnson** [comment]
I figured it out.  The latest PR broke how the versioning works.  I thought the line looked wrong but I wasn't sure.  The testing set up is a bit unfortunate right now.

**Comment by favoretti** [comment]
That script was broken forever, and never worked on anything but a windows machine. I've used to contribute to this extension when it was owned by Charles, but he gave us access to his ADO pipelines back then, so I never bothered to look into it. Thought I'd "fix" it, but apparently made it worse? :) Want me to revert that part in particular?

**Comment by favoretti** [comment]
Ahh, nevermind, I see you already did. Thank you and sorry! :)

**Comment by jason-johnson** [comment]
No problem.  It's fairly convoluted but the only way to offer different versions of an ADO extension is to package each version in the final package.  And without that we would have to have a much stricter release cycle that the current level of contribution just can't support.  I suppose the best thing would be to document how this all works. :)

**Comment by jaalexan1983** [comment]
@jason-johnson thanks for the update, are you able to post on here once the final package has been pushed to the market place.  Thanks in advance.

**Comment by jason-johnson** [comment]
Yes, if I can ever manage to get the tests to succeed (required to public production version to marketplace) I will post here

**Comment by favoretti** [comment]
Let me know if I can help. I recall some of what the pipelines looked like still.

**Comment by jason-johnson** [comment]
I managed to fix everything but two tests (I think they are simple permission issues).  If you'd like, I can give you permissions on the new ADO pipelines.  I just need to know what user to add.  If you don't want to write it here then we can figure out a way to communicate it somewhere else.

**Comment by favoretti** [comment]
I think you can find me on MSFT teams as a guest (Vladimir Lazarenko). I can't search for folks there, but I should be findable for you.

**Comment by jason-johnson** [comment]
I've almost got the tests working (which would allow the production version to release) but my ADO is causing all sorts of issues.  I'm considering moving most of the build completely away from there because I get so many issues from it.

**Comment by fardarter** [comment]
> I've almost got the tests working (which would allow the production version to release) but my ADO is causing all sorts of issues. I'm considering moving most of the build completely away from there because I get so many issues from it.

What issues are you facing?

**Comment by ag-advania** [comment]
I think this issue should not be closed until the Marketplace version get updated

**Comment by favoretti** [comment]
That's how GitHub works. Once PR is merged - issue is autoclosed.

**Comment by fardarter** [comment]
@favoretti It would make sense to reopen then. @jason-johnson perhaps if you can articulate some of the issues you're facing you may get some help?

**Comment by jason-johnson** [comment]
Ok, I've added you to ADO.  I had ran out of space on the ADO organization so no further pipelines would run.  I deleted most of the runs but it can take up to 48 hours to actually unlock again.  That is all fixed and all that remains is getting one more failing test to pass.  We have switched completely to Workflow Identity Federation so I've modified the tests and there is one left to fix.

**Comment by fardarter** [comment]
@jason-johnson I've not seen an invite of any kind. Space as is artifacts?

Where are the pipeline definitions?

**Comment by favoretti** [comment]
@jason-johnson I see that it fails with self-configured backend, but since you're using a securefile for it - I can't see what's in it.. I guess it's merely a question of wrong creds in the file?

**Comment by jason-johnson** [comment]
> @jason-johnson I've not seen an invite of any kind. Space as is artifacts?
> 
> Where are the pipeline definitions?

Oh sorry, it was favoretti who was invited.  He used to be on the project and didn't get added to the new setup.  The pipeline definitions are all in the repository.  The GitHub ones are in the normal place and the ADO stuff is under the pipeline folder.

**Comment by jason-johnson** [comment]
> @jason-johnson I see that it fails with self-configured backend, but since you're using a securefile for it - I can't see what's in it.. I guess it's merely a question of wrong creds in the file?

The file is:

```
use_oidc             = true
resource_group_name  = "rg-trfrm-rc-chn-jbj"
storage_account_name = "sttrfrmrcchnjbj"
container_name       = "azure-pipelines-terraform"
key                  = "self_configured.tfstate"
client_id            = "<redacted>"
subscription_id      = "<redacted>"
tenant_id            = "<redacted>" 
```

I'm not sure if it is incorrect configuration for Workflow Identity Federation, it is setup how the terraform documentation says it should be.  It's complaining that `az login` needs to be run and I'm not sure that's true.

**Comment by fardarter** [comment]
@jason-johnson Could you maybe bost the error?

**Comment by fardarter** [comment]
Fwiw, this is what I use with WIF on my terraform:
  
  ```tf
terraform {
    backend "azurerm" {
          resource_group_name  = "name"
          storage_account_name = "name"
          container_name       = "name"
          key                  = "key"
          use_azuread_auth     = true
          use_oidc             = true
      }
}

provider "azurerm" {
    storage_use_azuread             = true
}
```

**Comment by jason-johnson** [comment]
```
/opt/hostedtoolcache/terraform/1.9.8/x64/terraform init -backend-config=/home/vsts/work/_temp/backend_rc.vars
Initializing the backend...
╷
│ Error: Error building ARM Config: obtain subscription(7a2964f4-70b7-4541-aa12-7a134b79e105) from Azure CLI: parsing json result from the Azure CLI: waiting for the Azure CLI: exit status 1: ERROR: Please run 'az login' to setup account.
```

This is a test designed to test secure files and putting all the configuration you showed in the files instead of in the terraform configuration.  Also note that the service connection we use in ADO is only enabled over Workflow Identity Federation.  There is no secret or anything.

**Comment by fardarter** [comment]
@jason-johnson not missing `environmentServiceName`?

**Comment by jason-johnson** [comment]
That was never used for this test.  It has `backendType` set to `self-configured` so there is no service connection.

**Comment by fardarter** [comment]
@jason-johnson How do you acquire your first token?

**Comment by jason-johnson** [comment]
Honestly I'm not sure how this test was working before, maybe it can only be done when using secrets.  I suppose I could just turn it off to get the package published.

**Comment by jason-johnson** [comment]
@jaredfholgate would you happen to know if the Workflow Identity Federation authentication method can be used without "az login"?  If that's not enough information hopefully there is enough in this chat, the build logs and so on.  Or we can do a call to discuss.

**Comment by jason-johnson** [comment]
@favoretti @fardarter @oguzkilcan Version 1.1.2 has been released.  This means the `@1` and `@2` versions should work.  Please let me know if there is still a warning. I may need to patch or drop the `@0` version.

**Comment by jaredfholgate** [comment]
> @jaredfholgate would you happen to know if the Workflow Identity Federation authentication method can be used without "az login"? If that's not enough information hopefully there is enough in this chat, the build logs and so on. Or we can do a call to discuss.

Yes, it works with env vars. You need to supply the id token though, so it won't work with just a static backend file like the one shown above since the id token is per run. There is some magic in the provider for GitHub that allows you to just specifiy `use_oidc`, but that is implemented for Azure DevOps (yet). For Azure DevOps, you must supply `oidc_token` too, most likely via the env var: https://developer.hashicorp.com/terraform/language/backend/azurerm#oidc_token

**Comment by catt231** [comment]
> [@favoretti](https://github.com/favoretti) [@fardarter](https://github.com/fardarter) [@oguzkilcan](https://github.com/oguzkilcan) Version 1.1.2 has been released. This means the `@1` and `@2` versions should work. Please let me know if there is still a warning. I may need to patch or drop the `@0` version.

I think I'm missing something? I've updated to @1 but I'm still seeing the error
```
Download all required tasks.
Downloading task: TerraformInstaller (1.0.6)
##[warning]Task 'Terraform Installer' version 1 (TerraformInstaller@1) is dependent on a Node version (10) that is end-of-life. Contact the extension owner for an updated version of the task. Task maintainers should review Node upgrade guidance: https://aka.ms/node-runner-guidance
Downloading task: TerraformCLI (1.0.6)
##[warning]Task 'Terraform CLI' version 1 (TerraformCLI@1) is dependent on a Node version (10) that is end-of-life. Contact the extension owner for an updated version of the task. Task maintainers should review Node upgrade guidance: https://aka.ms/node-runner-guidance
```
Any ideas?

**Comment by jason-johnson** [comment]
Are you certain this is the installer from this project?  If so, can you make sure you've installed the latest version of the extension?
